Yaml support (#3)
dbmi-bgm/portal-pipeline-utils: docs/deploy_pipeline.rst

.. _deploy_pipeline:

======================================
Deploy Pipelines to a CGAP Environment
======================================

This document describes how to deploy pipelines to a target CGAP environment.

> Although it's possible to run the deployment from a local machine, we highly recommend using an EC2 machine.

Reviewer: Does this EC2 have to be within the same account as the portal to work successfully? I'd hope not, but we may want to test this if we haven't already.

Author: no it does not, just need credentials set up for the right account. I am using my wolf machine to deploy everywhere else

Setup an EC2 Machine
====================

This step may be skipped if you have an EC2 already set up.

We recommend using the following configuration:

  * AMI: Use a linux distribution (64-bit, x86)
  * Instance Type: t3.large or higher
  * Storage: 50+GB in main volume

Install Docker
==============

The deployment code will try to trigger remote codebuild jobs to build and push the Docker containers implemented for the pipelines directly in AWS.

> However, if no builder has been setup and it's available, the code will try to run a local build using Docker.

Reviewer: ```suggestion
However, if no builder has been setup, the code will try to run a local build using Docker, if available.
```

Running a local build requires having a Docker application running on the machine.
To install Docker in a EC2 machine, refer to the following instructions based on an Amazon Linux AMI:

> Update packages: ``sudo yum update -y``

Reviewer: ```suggestion
Update packages:
``sudo yum update -y``
```
For consistency with following lines. Also, consider placing all these lines of code within one block (with the descriptions as comments or something) so that they can be copy-pasted by users if need be.

Install the Docker Engine package:
``sudo yum install docker``

Start the docker service:
``sudo service docker start``

Ensure Docker is installed correctly and has the proper permissions
by running a test command:
``docker run hello-world``

More information on how to setup Docker can be found in the
`AWS Documentation <https://docs.aws.amazon.com/AmazonECS/latest/developerguide/docker-basics.html>`_.

We now need to install the ``pipeline_utils`` software to deploy the pipeline components.

Install pipeline_utils
======================

The software is python based.
To install the software and the required packages it is recommended to

> use a fresh virtual environment running Python 3.7 or 3.8.

Reviewer: For future proofing this documentation, consider dropping the references to the python versions here and pointing to `pyproject.toml` or something like that.

> use a fresh virtual environment running Python 3.7 or 3.8.

Reviewer: ```suggestion
To install the software and the required packages, we recommend
using a fresh virtual environment.
Please refer to ``pyproject.toml`` for the supported Python version.
```

We recommend using pyenv to manage virtual environments.
Instructions for installing and using
pyenv can be found `here <https://realpython.com/intro-to-pyenv/>`_.

Once the virtual environment is set up and activated, we can proceed to :ref:`install <install>` cgap-pipeline-utils software.

.. code-block:: bash

  # Install from source
  git clone https://github.com/dbmi-bgm/cgap-pipeline-utils.git
  cd cgap-pipeline-utils
  make configure
  make update
  make build
  cd ..

  # Install from pypi
  pip install cgap-pipeline-utils

To check that the software is correctly installed, try to run ``pipeline_utils``.

> If this fail with a bash "command not found" error, try ``poetry run pipeline_utils`` instead.

Reviewer: I don't believe you have poetry as a dependency in this package, so this is likely to fail if following from scratch. I'd recommend dropping this line or adding some context here about using poetry/ensuring it is available.

Author: right, the pypi doesn't come with poetry

Set Up Credentials and Environmental Variables
==============================================

AWS Auth Credentials
--------------------

To deploy pipelines components in a specific AWS account,
we need to setup the following environmental variables to authenticate to the account.

.. code-block:: bash

  export AWS_ACCOUNT_NUMBER=
  export TIBANNA_AWS_REGION=
  export GLOBAL_ENV_BUCKET=
  export S3_ENCRYPT_KEY=

  export AWS_ACCESS_KEY_ID=
  export AWS_SECRET_ACCESS_KEY=

  # Optional, depending on the account
  export S3_ENCRYPT_KEY_ID=
  export AWS_SESSION_TOKEN=

**Tips**:

> 

Reviewer: Where can one find `TIBANNA_AWS_REGION`?

* *GLOBAL_ENV_BUCKET* can be found in AWS S3. After logging into the AWS

>   console and navigating to the S3 page, this variable is the name of the bucket that ends in "-envs".

Reviewer: This is also in the secrets manager, and that's probably the best route to find it.

* *S3_ENCRYPT_KEY* and *S3_ENCRYPT_KEY_ID* can be found in the AWS Secrets Manager.
* *AWS_SESSION_TOKEN* is used by some single sign-on platforms for managing
  credentials but may not be required otherwise.

Portal Credentials
------------------

We also need to setup credentials to authenticate to the portal database to push some of the portal components.
These credentials need to be stored as a keypair file as described
`here <https://github.com/dbmi-bgm/cgap-portal/blob/master/docs/public/help/access_keys.md>`_.

The default path used by the code to locate this file is ``~/.cgap-keys.json``.

> However, it is possible to specify a different keypair file throug the command line, if desired.

Reviewer: ```suggestion
However, it is possible to specify a different keypair file through a command line argument, if desired.
```

Target Account Information
--------------------------

Finally we need to setup the information to identify the target environment to use for the deployment.

.. code-block:: bash

  # Set the name of the target environment
  #   e.g., cgap-wolf

>   export ENV_NAME=

Reviewer: This is within the secrets manager as well, so you probably don't need to set it. You may want to automate the setting of these environmental variables to increase ease of use here. You can probably make good use of some of the env utils in dcic-utils made by back-end; I imagine they've solved some of this already.

  # Set the bucket used to store the worklow description files
  #   e.g., cgap-biotest-main-application-tibanna-cwls
  export WFL_BUCKET=

  # Set the path to the keypair file with the portal credential
  #   e.g., /.cgap-keys.json

>   export KEYDICTS_JSON=

Reviewer: This is optional, right, since it will default to the typical location? If so, may want to indicate that.

Also, you need to know which key to use in the keys file. I don't believe that's being handled right now.

Also, the usual location is `~/.cgap-keys.json` not `/.cgap-keys.json`.

  # Set up project and institution
  #   Project and institution need to correspond to metadata present on the portal
  #   e.g., cgap-core and hms-dbmi
  export PROJECT=

>   export INSTITUTION=

Reviewer: These should probably be clarified a little. I imagine any identifying information for the objects would work, so either [one of these](https://github.com/dbmi-bgm/cgap-portal/blob/master/src/encoded/schemas/project.json#L9-L14) for a Project and [one of these](https://github.com/dbmi-bgm/cgap-portal/blob/master/src/encoded/schemas/institution.json#L10-L15) for an Institution, as well as their UUIDs.

  # If running sentieon code,
  #   specify the address for the server that validate the software license
  #   e.g., 0.0.0.0
  export SENTIEON_LICENSE=

**Tips and Options:**

* *ENV_NAME* can be found in the portal health page under ``Namespace``.

> * *WFL_BUCKET* can be found in the portal health page under ``Tibanna CWLs Bucket``.

Reviewer: You can automate these pretty easily then if you have the keys. I'd go that route to ease the burden of setting variables manually.

> * *WFL_BUCKET* can be found in the portal health page under ``Tibanna CWLs Bucket``.

Reviewer: Is this only the case for CWLs, or can WDLs go here as well?

Running the Deployment
======================

> The following code will use ``pipeline_deploy`` command to deploy all the components from the repositories specified

Reviewer: ```suggestion
The following code will use the ``pipeline_deploy`` command to deploy all the components from the repositories specified
```

by the ``--repos`` argument.

.. code-block:: bash

  pipeline_utils pipeline_deploy \
    --ff-env ${ENV_NAME} \
    --keydicts-json ${KEYDICTS_JSON} \
    --wfl-bucket ${WFL_BUCKET} \
    --account ${AWS_ACCOUNT_NUMBER} \
    --region ${TIBANNA_AWS_REGION} \
    --project ${PROJECT} \
    --institution ${INSTITUTION} \
    --sentieon-server ${SENTIEON_LICENSE} \
    --post-software \
    --post-file-format \
    --post-file-reference \
    --post-workflow \
    --post-metaworkflow \
    --post-wfl \
    --post-ecr \
    --repos REPO [REPO ...]

It is possible to add flags to run the command in various debug modes, to validate the objects and test the pipeline implementation without running a real deployment.
For more details on the command line arguments refer to the documentation for the :ref:`pipeline_deploy <pipeline_deploy>` command.

> An important argument is ``--branch``, this argument specifies the branch to check out for cgap-pipeline-main to build ECR through codebuild.

Reviewer: ```suggestion
An important argument is ``--branch``; this argument specifies the branch to check out for cgap-pipeline-main to build ECR through codebuild.
```

> The default is set to the main branch. The ``--local-build`` flag will prevent the code to try using codebuild and force a local build using Docker instead.

Reviewer: ```suggestion
The default is set to the main branch. The ``--local-build`` flag will prevent the code from using codebuild and force a local build with Docker instead.
```

> *Note: we are working to enable more builder and a selection with a command line argument for which builder to use to deploy modules from different repositories through codebuild.*

Reviewer: ```suggestion
*Note: we are working to enable more builders with a command line argument for which builder to use to deploy modules from different repositories through codebuild.*
```

Deploying CGAP Pipelines
========================

CGAP pipelines are released as a complete package with a customized set up for automated deployment to the desired environment.
To deploy the pipelines run the following steps:

1. Clone the main pipeline repository.
The submodules will be empty and set to the current commits saved for the main branch.

.. code-block:: bash

  git clone https://github.com/dbmi-bgm/cgap-pipeline-main.git

2. Check out the desired version.
This will set the submodules to the commits saved for that pipeline release.

.. code-block:: bash

  git checkout <version>

3. Download the content for each submodule.
The submodules will be set in detached state on their current commit.

.. code-block:: bash

  make pull

4. Build pipeline_utils (optional).
This will build from source the latest version linked for the current release.

.. code-block:: bash

  make configure
  make update
  make build

5. Set up the auth credentials as described above.

> 6. Set the target account information in the ``.env`` file.

Reviewer: This hasn't been mentioned previously and may warrant some explanation or a pointer to said explanation.

7. Test the deployment using the base module only.

.. code-block:: bash

  make deploy-base

8. Deploy all the other modules.

.. code-block:: bash

>   makde deploy-all

Reviewer: ```suggestion
  make deploy-all
```

.. _deploy_pipeline:

======================================
Deploy Pipelines to a CGAP Environment
======================================

This document describes how to deploy pipelines to a target CGAP environment.
Although it's possible to run the deployment from a local machine, we highly recommend using an EC2 machine.

Setup an EC2 Machine
====================

This step may be skipped if you have an EC2 already set up.

We recommend using the following configuration:

  * AMI: Use a linux distribution (64-bit, x86)
  * Instance Type: t3.large or higher
  * Storage: 50+GB in main volume

Install Docker
==============

The deployment code will try to trigger remote codebuild jobs to build and push the Docker containers implemented for the pipelines directly in AWS.

> However, if no builder has been setup and it's available, it is possible to run a local build using Docker by passing the flag ``--local-build`` to the deployment command.

Author: ```suggestion
However, if no builder has been setup, it is possible to run a local build using Docker by passing the flag ``--local-build`` to the deployment command.
```

by running a test command:
``docker run hello-world``

More information on how to setup Docker can be found in the
`AWS Documentation <https://docs.aws.amazon.com/AmazonECS/latest/developerguide/docker-basics.html>`_.

We now need to install the ``pipeline_utils`` software to deploy the pipeline components.

Install pipeline_utils
======================

The software is python based.
To install the software and the required packages, we recommend
using a fresh virtual environment.
Please refer to ``pyproject.toml`` for the supported Python version.

We recommend using pyenv to manage virtual environments.
Instructions for installing and using
pyenv can be found `here <https://realpython.com/intro-to-pyenv/>`_.

Once the virtual environment is set up and activated, we can proceed to :ref:`install <install>` cgap-pipeline-utils software.

.. code-block:: bash

  # Install from source
  git clone https://github.com/dbmi-bgm/cgap-pipeline-utils.git
  cd cgap-pipeline-utils
  make configure
  make update
  make build
  cd ..

  # Install from pypi
  pip install cgap-pipeline-utils

To check that the software is correctly installed, try to run ``pipeline_utils``.

> If this fail with a bash "command not found" error, try ``poetry run pipeline_utils`` instead.

Author: ```suggestion
If installed from source, this command may fail with a bash "command not found" error, try ``poetry run pipeline_utils`` instead.
```

Set Up Credentials and Environmental Variables
==============================================

AWS Auth Credentials
--------------------

To deploy pipelines components in a specific AWS account,
we need to setup the following environmental variables to authenticate to the account.

.. code-block:: bash

  export AWS_ACCOUNT_NUMBER=
  export TIBANNA_AWS_REGION=
  export GLOBAL_ENV_BUCKET=
  export S3_ENCRYPT_KEY=

  export AWS_ACCESS_KEY_ID=
  export AWS_SECRET_ACCESS_KEY=

  # Optional, depending on the account
  export S3_ENCRYPT_KEY_ID=
  export AWS_SESSION_TOKEN=

**Tips**:

> * *GLOBAL_ENV_BUCKET* can be found in AWS S3. After logging into the AWS

Author: ```suggestion
* *GLOBAL_ENV_BUCKET* can be found in the AWS Secrets Manager.
```

>   console and navigating to the S3 page, this variable is the name of the bucket that ends in "-envs".

Author: ```suggestion
```

* *S3_ENCRYPT_KEY* and *S3_ENCRYPT_KEY_ID* can be found in the AWS Secrets Manager.
* *AWS_SESSION_TOKEN* is used by some single sign-on platforms for managing
  credentials but may not be required otherwise.

> * *TIBANNA_AWS_REGION* is the region for the AWS account.

Author: ```suggestion
* *TIBANNA_AWS_REGION* is the main region for the AWS account.
```
